### ISSUE TITLE
BUG-Service deleted from service group while merging duplicate services to the upper level device group

### ISSUE THREAD

**Issue by gasin1** [PaloAltoNetworks/pan-os-php#524]
## The script deletes service that starts with **tcp_** from a service group while merging other service (in same service group) to upper level.

I was validating config logs and found that before change service_group1 contains members tcp_5670 and tcp-9898 but after  

change shows Service_groups1 contains only member tcp_9898 (merged with the shared)

 There is no duplicity for tcp_5670 in upper level.
 **[the bug is available, if same service object name and value: tcp_5670 with value 5670 is available at Lower_level_DG and at parent/shared DG level]** (comment by swaschkut)

The config logs in Panorama show as below: 

The script used: php pan-os-php.php type=service-merger in=api://panorama.domain.com location=Lower_level_DG  allowMergingWithUpperLevel 

before-change-detail 
Service_object1  { members [ tcp_5670 tcp-9898 ]; } 

after-change-detail
Service_object1 { members [ tcp_9898 ]; }

## Steps to reproduce

1. In a lab environment, create a service object with name tcp-9898 and tcp_5670 in lower level device group.
2. Create service tcp_9898 and tcp_5670 in upper level (Parent) device group.
3. Now create a service group (service_group1) and add members tcp-9898 and tcp_5670 in the service group.
4. Now the services are being used in the service group in lower level.
5. run the script : php pan-os-php.php type=service-merger in=api://panorama.domain.com location=Lower_level_DG  allowMergingWithUpperLevel 
6. After merging completes, check the service group in lower level device group to see if the ´service group has both the members.

## Your Environment

- Version used: Panorama software version: 9.1.12-h3
- Environment name and version (e.g. Chrome Version 99.0.4844.51: ):
- Operating System and version (OS version on which Panosphp is running: ubuntu "20.04.1 LTS ):
- Link to your project:

**Comment by swaschkut** [comment]
I just tried what you explained and found based on the "bug".

original config:

pan-os-php type=service in=/tmp/bug.xml 'filter=(name eq Service_object1) or (name regex /5670/) or (name regex /9898/)' location=any 
==================================================================

* processing store 'PanoramaConf: / ServiceStore:services' that holds 1 objects

   - object 'tcp_9898' passing through Action='display'

     * Service 'tcp_9898'     value: 'tcp/9898'

 

 

* objects processed in DG/Vsys '' : 1

 

 

* processing store 'PanoramaConf: / DeviceGroup:parent / ServiceStore:services' that holds 0 objects

 

* objects processed in DG/Vsys 'parent' : 0

 

 

* processing store 'PanoramaConf: / DeviceGroup:child / ServiceStore:services' that holds 10 objects

   - object 'tcp-9898' passing through Action='display'

     * Service 'tcp-9898'     value: 'tcp/9898'

 

   - object 'tcp-5670' passing through Action='display'

     * Service 'tcp-5670'     value: 'tcp/5670'

 

   - object 'Service_object1' passing through Action='display'

     * ServiceGroup 'Service_object1'

          - tcp-5670    value: 'tcp/5670'

          - tcp-9898    value: 'tcp/9898'

==================================================================


run command:

pan-os-php type=service-merger in=/tmp/bug.xml out=/tmp/bugfix.xml location=child allowmergingwithupperlevel

==================================================================
final config:

 pan-os-php type=service in=/tmp/bugfix.xml 'filter=(name eq Service_object1) or (name regex /5670/) or (name regex /9898/)' location=any

 

* processing store 'PanoramaConf: / ServiceStore:services' that holds 1 objects

   - object 'tcp_9898' passing through Action='display'

     * Service 'tcp_9898'     value: 'tcp/9898'

 

 

* objects processed in DG/Vsys '' : 1

 

 

* processing store 'PanoramaConf: / DeviceGroup:parent / ServiceStore:services' that holds 0 objects

 

* objects processed in DG/Vsys 'parent' : 0

 

 

* processing store 'PanoramaConf: / DeviceGroup:child / ServiceStore:services' that holds 9 objects

   - object 'tcp-5670' passing through Action='display'

     * Service 'tcp-5670'     value: 'tcp/5670'

 

   - object 'Service_object1' passing through Action='display'

     * ServiceGroup 'Service_object1'

          - tcp-5670    value: 'tcp/5670'

          - tcp_9898    value: 'tcp/9898'

==================================================================


The same correct merging I am seeing if I am doing API mode:
pan-os-php type=service-merger in=api://PANORAMA-IP location=child allowmergingwithupperlevel

There is no bug in the actual PAN-OS-PHP version available:


 pan-os-php version

 - PAN-OS-PHP version: 2.0.32 [UNIX]

 - /Users/swaschkut/Documents/PAN-scripting/pan-os-php/utils

 - PHP version: 8.1.0

**Comment by gasin1** [comment]
I dont know why its not showing all the information that I have motioned in the above bug: 

**Expected behavior**: It should replace the service from the upper level (shared device)  to lower level where ever it is being used but not delete any service in the same service group where it has replaced the other service with upper DG.

**Current behavior:** In a service group for example Service-group1 there were two members, **tcp_5670** and **tcp-9898**, it replaces the **tcp-9898** with tcp_9898 from upper level service but at the same time it also deletes the other member (tcp_5670) . So after merge the service_group1 only has one member **tcp_9898**.
Point to be notes that the service which got deleted from the service group (tcp_5670) has the exactly same name service (tcp_5670) in upper level device group as well.

**Possible solution**: 
It should only replace the member which is merged with the upper level service in service group without deleting other members in the same service group.
Right now I dont have any idea for alternative solution.

**How has this issue affected you? What are you trying to accomplish?**
I am trying to merge all the local services to the standard services in upper level shared device group to optimize the objects as well as lower the configuration size of the panorama. 
It will also help the customer to automate object creation through alsgosec easier and quicker.

**Your Environment:** 
1. Shared DG (upper level) has 3000 services.
2. Lower level DG has 126 services that need to be merged with upper level.
3. Lower level DG has 25 services which has exactly same name as upper level service, for example service name tcp_5670 is in both DGs.
4. All other 101 services have different name for example in upper level service has tcp_9898 but in lower level it is tcp-9898.

-Version used: Panorama software version: 9.1.12-h3
- Environment name and version (e.g. Chrome Version 99.0.4844.51: ):
- Operating System and version (OS version on which Panosphp is running: ubuntu "20.04.1 LTS ):

**Comment by swaschkut** [comment]
based on the testing, And the knowledge that this exact command:
pan-os-php type=service-merger location=any allowmergingwithupperlevel

was used last week by a big customer, with. more objects as you have,
I can not reproduce your issue.

maybe you can write me in person:
swaschkut (at) paloaltonetworks.com

to clarify things and have a Zoom meeting

**Comment by gasin1** [comment]
Did you have tcp_5670 (not tcp-5670 ) in the shared as well tcp_5670 in child DG?

Is is possible to use Script : php pan-os-php.php type=service-merger in=api://itseelm-bb4289.ikea.com location=DC05-Mgt01 allowMergingWithUpperLevel         (without regex)?

**Comment by swaschkut** [comment]
I am happy to demonstrate, that your object was NOT removed by PAN-OS-PHP.
Please reach out to swaschkut (at) paloaltonetworks.com to schedule a Zoom meeting

**Comment by gasin1** [comment]
That would be great, I have send you an email, we can schedule a zoom meeting to discuss. Thanks

**Comment by swaschkut** [comment]
problem can be still not reproduced;
assumption - 
related servicegroup has duplicate members, which is possible displayed as warning by pan-os-php;
these duplicate members can not be added via PAN-OS GUI, always seen by using of "load config partial" from PAN-OS CLI

waiting for feedback about related warning messages

**Comment by swaschkut** [comment]
bug fixed with PR:
#526 526


Please validate

**Comment by swaschkut** [comment]
bugfix validated and confirmed by customer that problem is solved

**Comment by gasin1** [comment]
I have just validated all the service groups in the local device group and it seems issue has been resolved. 

No missing members from the service groups after running the script.

To conclude, issue which was specific to services which has the exact name in local and shared device groups were merged with shared services and get removed from the service group in local device group when the service-merger allowMergingWithUpperLevel script was pushed.

Thank you for looking into the issue and fixing it with priority.
